### PR TITLE
Added trait, MDSLightPermuation

### PR DESCRIPTION
**Pull request by SyxtonPrime**
This trait encodes the properties that a transformation needs to satisfy in order to be used as the external_linear_layer in the external (full) rounds of Poseidon2.

Currently this trait is not used but in a future PR we will modify Poseidon2 to no longer hardcode the external_linear_layers.

This will allow us to optimise this external layer for specific fields/architectures.

### CODE REVIEW

File: poseidon2/src/matrix.rs [hunk 1]
```diff
@@ -1,23 +1,23 @@
 use p3_field::{AbstractField, PrimeField};
+use p3_mds::MdsPermutation;
 use p3_symmetric::Permutation;
 
 extern crate alloc;
 
-// The t x t matrix M_E := circ(2M_4, M_4, ..., M_4), where M_4 is the 4 x 4 matrix
+// For the external layers we use a matrix of the form circ(2M_4, M_4, ..., M_4)
+// Where M_4 is a 4 x 4 MDS matrix. This leads to a permutation which has slightly weaker properties to MDS
+pub trait MDSLightPermutation<T: Clone, const WIDTH: usize>: Permutation<[T; WIDTH]> {}
```
Comment (dlubarov) [inline poseidon2/src/matrix.rs:9]: Couple nits,

- I would use `///` so the comment shows up in docs, IDE hints etc.
- There's a [convention](https://rust-lang.github.io/api-guidelines/naming.html) for camel case acronyms, like `MdsLightPermutation`

**Review by unzvfu — APPROVED** [review]
Looks good, modulo some minor nits.

File: poseidon2/src/matrix.rs [hunk 2]
```diff
@@ -57,53 +57,98 @@ where
...
     }
 }
+
+#[derive(Default, Clone)]
+pub struct Poseidon2ExternalMatrixHL;
+
+impl<AF, const WIDTH: usize> Permutation<[AF; WIDTH]> for Poseidon2ExternalMatrixHL
+where
+    AF: AbstractField,
+    AF::F: PrimeField,
```
[the remark this answers is not in the document]
Comment (SyxtonPrime) [reply poseidon2/src/matrix.rs:142]: That's true, I guess this and the next change (impl of Permutation and MdsLightPermutation) are laying the groundwork for the next PR where I modify Poseidon2 to require input of an external_linear_layer which satisfies the trait MdsLightPermutation.

Should I take these changes out of this PR and move them into the next one?

File: poseidon2/src/matrix.rs [hunk 3]
```diff
@@ -1,23 +1,23 @@
...
-// Multiply a 4-element vector x by M_4, in place.
-// This uses the formula from the start of Appendix B, with multiplications unrolled into additions.
-fn apply_m_4<AF>(x: &mut [AF])
+// This uses the formula from the start of Appendix B in the Poseidon2 paper, with multiplications unrolled into additions.
+// It is also the matrix used by the Horizon Labs implementation.
+fn apply_hl_mat4<AF>(x: &mut [AF])
 where
     AF: AbstractField,
-    AF::F: PrimeField,
 {
```
[the remark this answers is not in the document]
Comment (SyxtonPrime) [reply poseidon2/src/matrix.rs:21]: Yeah in a future PR I the plan is to replace this input type by &mut [AF; 4]

File: poseidon2/src/matrix.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (unzvfu) [reply poseidon2/src/matrix.rs:138]: Yeah, I think that would be clearer, thanks.

**Review by unzvfu — APPROVED** [review]
Looks good, thanks!

File: poseidon2/src/matrix.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (SyxtonPrime) [reply poseidon2/src/matrix.rs:138]: Okay, I've moved the impl for MdsLightPermutation to the next PR.

Realized we need to keep the impl for Permutation as we use the .permute_mut method in lib.

**Pull request merged**